sys/fuchsia/test/job

```diff
@@ -0,0 +1,45 @@
+# Expect failure when creating a job from an invalid handle.
+
+zx_job_create(0x0, 0x0, &AUTO) # ZX_ERR_BAD_HANDLE
+
+# Create a valid job handle and several child jobs.
+
+r1 = syz_job_default()
+zx_job_create(r1, 0x0, &AUTO=<r2=>0x0)
+zx_job_create(r1, 0x0, &AUTO=<r3=>0x0)
+zx_job_create(r1, 0x0, &AUTO=<r4=>0x0)
+
+# With policy format ZX_JOB_POL_BASIC_V1, set policy ZX_POL_ACTION_ALLOW for condition ZX_POL_NEW_ANY.
+
+zx_job_set_policy$BASIC_V1(r2, 0x0, 0x0, &AUTO=[{0x3, 0x0}], 0x1)
+
+# With policy format ZX_JOB_POL_TIMER_SLACK, set a min slack of 256ns with a default mode of ZX_TIMER_SLACK_CENTER.
+
+zx_job_set_policy$TIMER_SLACK(r4, 0x0, 0x1, &AUTO=[{0x100, 0x0, "00000000"}], 0x1)
+
+# With policy format ZX_JOB_POL_BASIC_V2, set policy ZX_POL_ACTION_DENY for condition ZX_POL_NEW_ANY with flag ZX_POLICY_OVERRIDE_DENY.
+
+zx_job_set_policy$BASIC_V2(r3, 0x0, 0x01000000, &AUTO=[{0x3, 0x1, 0x1}], 0x1)
+
+# Setting the same policy again should succeed with either the ZX_JOB_POL_ABSOLUTE or ZX_JOB_POL_RELATIVE option.
+
+zx_job_set_policy$BASIC_V2(r3, 0x0, 0x01000000, &AUTO=[{0x3, 0x1, 0x1}], 0x1)
+
+# Setting a conflicting policy with the ZX_JOB_POL_RELATIVE option should succeed.
+
+zx_job_set_policy$BASIC_V2(r3, 0x0, 0x01000000, &AUTO=[{0x3, 0x0, 0x1}], 0x1)
+
+# Setting a conflicting policy with the ZX_JOB_POL_ABSOLUTE option should fail.
+
+zx_job_set_policy$BASIC_V2(r3, 0x1, 0x01000000, &AUTO=[{0x3, 0x0, 0x1}], 0x1) # ZX_ERR_ALREADY_EXISTS
+
+# Create a grandchild job. Expect failure when setting a previously allowed policy once a job has a child job.
+
+zx_job_create(r3, 0x0, &AUTO=<r5=>0x0)
+zx_job_set_policy$BASIC_V2(r3, 0x0, 0x01000000, &AUTO=[{0x3, 0x1, 0x1}], 0x1) # ZX_ERR_BAD_STATE
+
+# Create a process and set it as critical for a job.
+
+r6 = syz_job_default()
+r7 = syz_process_self()
```

Comment:
Something I puzzled over here, which maybe the reviewers will have some insight into:

According to https://fuchsia.dev/fuchsia-src/reference/syscalls/job_set_critical, the process should be a descendant of the job in order for zx_job_set_critical to succeed. I originally tried to create a child process like this:

zx_process_create(r6, "test_proc", 0x9, 0x0, &AUTO=<r7=>0x0, &AUTO)

...which I would expect to succeed, given examples like this where a job returned by `zx_job_default()` has the necessary rights to create a process: https://cs.opensource.google/fuchsia/fuchsia/+/main:zircon/system/utest/core/job/jobs.cc;l=58;drc=ed1095dcd018339e2d6eb7f223333d14a894a9b6

However, when included in the test, that line returns `ZX_ERR_ACCESS_DENIED`. 

The current version of the test passes, but I'm a little surprised that it does, given that there's no parent/child relationship (that I know of) between the job and the process passed to `zx_job_set_critical`.

Reply:
It looks like the zircon job test you're looking at works because it only runs in a special environment: https://cs.opensource.google/fuchsia/fuchsia/+/main:zircon/system/utest/core/BUILD.gn;l=73;drc=9fb8ac62d0e2baca0ee387502c8aa022ec5caa6c

Note the last paragraph in https://fuchsia.dev/fuchsia-src/concepts/process/process_creation?hl=en describing how this direct syscall is never really supported from within normal components unless you do special things like in `src/lib/process_builder/root_job_test_runner/main.rs`. Fetching the root job like that is maybe a possibility, but it might be cleaner to use this `job_policy_create_raw_processes` flag I came across: https://cs.opensource.google/fuchsia/fuchsia/+/main:docs/concepts/components/v2/elf_runner.md;l=136;drc=ccea478e1b7de941993b65ea6083de1421f3d450

`syz-executor` is a little unusual in that it's built as a shell package (https://cs.opensource.google/fuchsia/fuchsia/+/main:src/testing/fuzzing/syzkaller/BUILD.gn;l=27;drc=8c4e72633668edcfd83d99cb5a98ad8907d4c25d) which I think in the past may have granted it some higher privileges, but as far as I can tell it just means that at most it shares a namespace with sshd (https://cs.opensource.google/fuchsia/fuchsia/+/main:src/developer/sshd-host/meta/sshd-host.cml;drc=5eb51f914054441f4860aac7bc47e9e7db620891).  Therefore, it seems likely that the default job being passed is still of insufficient privilege, as you are seeing. (Interestingly, here's where the elf_runner removes the power: https://cs.opensource.google/fuchsia/fuchsia/+/main:src/sys/component_manager/src/elf_runner/mod.rs;l=335;drc=bb71f23bb1084d0c0a8d7e27fa597a4624852d20)

I haven't fully tracked down how exactly shell package binaries are spawned yet, though, so some parts of what I'm saying may be slightly incorrect.  (For reference, trampoline shims are generated by https://cs.opensource.google/fuchsia/fuchsia/+/main:src/lib/assembly/shell_commands/src/shell_commands.rs;l=78;drc=18757fe30a4b2438a196b9a62a3e674074fe76e1 which are picked up by   https://cs.opensource.google/fuchsia/fuchsia/+/main:sdk/lib/fdio/spawn.cc;l=216;drc=1dee12abd68777cfbc2e0f94c65f2e39259d2560).

Apologies for the rambling response, this was kind of a research exercise for me.

Reply:
Thank you, that was very helpful! It does look like `syz-executor`'s job policy is determined by `sshd-host`'s, and we'll probably need a different setup in order to fuzz some privileged syscalls. (E.g., as an experiment, I locally added `job_policy_create_raw_processes` to `sshd-host`'s manifest and allowlisted it in the component manager policy, and then was able to run a test like this:
```
zx_process_create(r3, &AUTO='test_proc', 0x9, 0x0, &AUTO=<r6=>0x0, &AUTO)
zx_job_set_critical(r3, 0x0, r6)
zx_handle_close(r6)
zx_job_create(r3, 0x0, &AUTO=<r8=>0x0) # ZX_ERR_BAD_STATE
```
...but that's clearly not the right solution.)

As for the potential false negative I was worried about with `zx_job_set_critical`, I realized that `syz_process_self()` just returns a handle to the `syz_executor` process, which is indeed a child of the default job, so there's no issue there. (Also, while experimenting with those local changes to `sshd-host`'s permissions, I did get the expected error when calling `zx_job_set_critical` with an unrelated job/process pair.) However, the test is incomplete in that we can't fully test the consequence of `zx_process_set_critical`. Ideally we'd include the second half of the 4-line program above, but first we need to be able to create subprocesses.

What do you think about committing this version for now, with a TODO for the loose end?

Reply:
Sounds like a great plan to me!  It's definitely worth following up on, not only for this test but because what you uncovered makes it clear we aren't properly covering `zx_process_create` either... and probably other things too.

P.S. It seems like the program language only supports constant values and not constant names as args, right?